packages/vite/src/node/optimizer/scan.ts

```diff
@@ -253,13 +253,24 @@ async function computeEntries(environment: ScanEnvironment) {
   if (explicitEntryPatterns) {
     entries = await globEntries(explicitEntryPatterns, environment)
   } else if (buildInput) {
-    const resolvePath = (p: string) => path.resolve(environment.config.root, p)
+    const resolveRollupInputPath = (p: string) => {
+      const resolvedPath = path.resolve(
+        environment.config.root,
+        p.replace(/^\//, ''),
+      )
+      if (fs.existsSync(resolvedPath)) return resolvedPath
+      for (const extension of environment.config.resolve.extensions) {
+        const pathWithExtension = `${resolvedPath}${extension}`
+        if (fs.existsSync(pathWithExtension)) return pathWithExtension
+      }
+      throw new Error('invalid rollupOptions.input value.')
+    }
```

Comment:
I made this change because discovery can start from `rollupOptions.input` values which can be paths such as:
https://github.com/vitejs/vite/blob/30975a8271bfc09f71b7009ab2fe45f9486c9baa/playground/environment-react-ssr/vite.config.ts#L47

The problem is that these paths can start with `/` and may also not include the file extension both things that break the path resolution here

I think this shows a slight issue in the fact that `optimizeDep.entries` and `rollupOptions.input` can be used in the same way here but they do accept different type of values (not a huge issue but I guess this can lead to confusion)

<details>

<summary>the leading `/` issue</summary>

Just to clarify the issue with the leading `/`, if an entry in `path.resolve` starts with `/` then the previous entries are thrown away (this is an expected behavior as per the [nodejs docs](https://nodejs.org/api/path.html#pathresolvepaths)):
![Screenshot 2024-10-18 at 10 43 03](https://github.com/user-attachments/assets/2468ae52-d9de-429b-a2c4-1abff968e724)

</details>

Reply:
I'll give that a try, thanks! 🙂

Reply:
`environment.pluginContainer.resolveId` works fine but if I provide a `rollupOptions.input` set to an incorrect value that doesn't start with `/` or `./` then the process hangs (instead of erroring)

e.g.
if I have
```js
  rollupOptions: {
    input: {
      index: 'zrc/entry-server',
    },
  },
```
the vite dev server hangs indefinitely

I'm trying to understand why 👀

Reply:
Ok I think I figured it out, if a path does not start with `/` or `./` then `vite:resolve` tries to resolve it as a bare import, meaning that [it goes through `tryOptimizedResolve`](https://github.com/vitejs/vite/blob/dbb260499f894d495bcff3dcdf5635d015a2f563/packages/vite/src/node/plugins/resolve.ts#L406) but here we're waiting for the scan to complete, so the process just hangs.

Basically with the setup above with these changes:
![Screenshot 2024-10-21 at 12 31 51](https://github.com/user-attachments/assets/509b0d0e-b88e-41e2-bad8-29b4a431f8b5)
I never encounter the second `debugger` statement

I am thinking of updating the path here, before passing it to `environment.pluginContainer.resolveId` to add a leading `/` in case the path doesn't start with `./`, `../` nor `/`

does that sound like a good solution @sapphi-red?

Reply:
Ah, I see. I think passing `scan: true` to indicate that it's for scan can fix it.
https://github.com/vitejs/vite/pull/18358/files#r1810071111

Reply:
yeah `scan: true` seems to work perfectly!

Sorry for getting the wrong result 😅  thanks! 😃